### PR TITLE
option,helm: Add a flag enable-k8s-networkpolicy

### PR DESCRIPTION
**Pull request by ChengyuanLiCY**
This flag is for cilium agent to watch K8s NetworkPolicy. By default the value is true. User can set "enable-k8s-networkpolicy = false" to disable it.

```release-note
option,helm: Add a flag to opt out from support for Kubernetes NetworkPolicy in Cilium
```

### CODE REVIEW

File: install/kubernetes/cilium/values.yaml
Comment (gandro) [inline install/kubernetes/cilium/values.yaml]: nit, language:

```suggestion
  # -- Enable support for K8s NetworkPolicy
```

File: install/kubernetes/cilium/values.yaml
Comment (gandro) [inline install/kubernetes/cilium/values.yaml]: nit:

```suggestion
k8sNetworkPolicy:
```

File: Documentation/cmdref/cilium-agent.md
Comment (gandro) [inline Documentation/cmdref/cilium-agent.md]: I think this doesn't match the current description.

I would also consider improving the language a bit, e.g. `Enables K8s NetworkPolicy support`, since the meaning of `watch` might not be clear to users

**Review by gandro — COMMENTED** [review]
Thanks! I am curious, what's the use case for this? Overall I think Helm and flags look fine, I left a bit of feedback regarding the naming and description.

**Comment by ChengyuanLiCY** [comment]
@gandro Thanks for the review!
> I am curious, what's the use case for this?
We are working on installing Cilium on existing kubernetes production clusters, we don't want to break anything in the production, so need to enable Cilium features little by little.

**Review by gandro — CHANGES_REQUESTED** [review]
The Go linter also failed with a legitimate error:

```
Error: /home/runner/work/cilium/cilium/src/github.com/cilium/cilium/pkg/k8s/watchers/watcher_test.go:196:3: cannot use &(fakeWatcherConfiguration literal) (value of type *fakeWatcherConfiguration) as WatcherConfiguration value in argument to NewK8sWatcher: *fakeWatcherConfiguration does not implement WatcherConfiguration (missing method K8sNetworkPolicyEnabled)
```

**Comment by gandro** [comment]
Doc lint and Travis failures are also legitimate:

Docs:
```
Running linter...
Validating documentation (syntax, spelling)...

Please fix the following spelling mistakes:
* Documentation/helm-values.rst:1376: (sNetworkPolicy)  k8sNetworkPolicy
* Documentation/helm-values.rst:1380: (sNetworkPolicy)  k8sNetworkPolicy.enabled
* Documentation/helm-values.rst:1376: (sNetworkPolicy)  k8sNetworkPolicy
* Documentation/helm-values.rst:1380: (sNetworkPolicy)  k8sNetworkPolicy.enabled

If the words are not misspelled, run:
Documentation/update-spelling_wordlist.sh sNetworkPolicy 
```

Travis (this is related to https://github.com/cilium/cilium/pull/23127#discussion_r1072098711 )
```
  CHECK  Documentation cmdref
diff --git a/Documentation/cmdref/cilium-agent.md b/Documentation/cmdref/cilium-agent.md
index dc92b89b..a8afb8d1 100644
--- a/Documentation/cmdref/cilium-agent.md
+++ b/Documentation/cmdref/cilium-agent.md
@@ -114,7 +114,7 @@ cilium-agent [flags]
       --enable-k8s-api-discovery                                Enable discovery of Kubernetes API groups and resources with the discovery API
       --enable-k8s-endpoint-slice                               Enables k8s EndpointSlice feature in Cilium if the k8s cluster supports it (default true)
       --enable-k8s-event-handover                               Enable k8s event handover to kvstore for improved scalability
-      --enable-k8s-networkpolicy                                Enable support for K8s NetworkPolicy
+      --enable-k8s-networkpolicy                                Enable the K8s NetworkPolicy (default true)
       --enable-k8s-terminating-endpoint                         Enable auto-detect of terminating endpoint condition (default true)
       --enable-l2-neigh-discovery                               Enables L2 neighbor discovery used by kube-proxy-replacement and IPsec (default true)
       --enable-l7-proxy                                         Enable L7 proxy for L7 policy enforcement (default true)

HINT: to fix this, run 'make -C Documentation update-cmdref'
```

**Comment by gandro** [comment]
@ChengyuanLiCY Please do not remove request for reviews. Those people are assigned for each code owner (e.g. docs, cli, k8s) to also give a review. Only once all code owners have approved this PR will it be merged.

**Comment by gandro** [comment]
There still is an issue with the Helm docs. Travis is failing with:

```

index b6bc96cf..8a1606b5 100644

--- a/Documentation/helm-values.rst

+++ b/Documentation/helm-values.rst

@@ -1373,12 +1373,8 @@

      - Configure Kubernetes specific configuration

      - object

      - ``{}``

-   * - k8sNetworkPolicy

-     - Enable support for K8s NetworkPolicy in cilium agent

-     - object

-     - ``{"enabled":false}``

    * - k8sNetworkPolicy.enabled

-     - Enable support for K8s NetworkPolicy in cilium agent

+     - Enable support for K8s NetworkPolicy

      - bool

      - ``true``

    * - k8sServiceHost

HINT: to fix this, run 'make -C Documentation update-helm-values'

make[1]: *** [Makefile:69: check] Error 1

make: *** [Makefile:622: postcheck] Error 2
make: *** Waiting for unfinished jobs....
The command "./.travis/build.sh" exited with 2.
```

Please run 'make -C Documentation update-helm-values' to fix this.

**Comment by tommyp1ckles** [comment]
Seems like a possible flake in the Travis tests

**Review by gandro — APPROVED** [review]

**Comment by tommyp1ckles** [comment]
/test-1.25-4.19

File: pkg/defaults/defaults.go
Comment (qmonnet) [inline pkg/defaults/defaults.go]: s/need/needs/, but then we don't need to give so many precisions (or to provide the default value again) here.
```suggestion
	// EnableK8sNetworkPolicy enables support for K8s NetworkPolicy.
```

File: daemon/cmd/daemon_main.go
Comment (qmonnet) [inline daemon/cmd/daemon_main.go]: Consistency with the other descriptions, plus "the" is probably not needed.
```suggestion
	flags.Bool(option.EnableK8sNetworkPolicy, defaults.EnableK8sNetworkPolicy, "Enable support for K8s NetworkPolicy")
```

File: pkg/k8s/utils/utils.go
Comment (qmonnet) [inline pkg/k8s/utils/utils.go]:
```suggestion
	// K8sNetworkPolicyEnabled returns true if cilium agent needs to support K8s NetworkPolicy
```

File: pkg/option/config.go
Comment (qmonnet) [inline pkg/option/config.go]:
```suggestion
	// EnableK8sNetworkPolicy enables support for K8s NetworkPolicy.
```

File: pkg/option/config.go
Comment (qmonnet) [inline pkg/option/config.go]:
```suggestion
	// EnableK8sNetworkPolicy enables support for K8s NetworkPolicy.
```

File: pkg/option/config.go
Comment (qmonnet) [inline pkg/option/config.go]: Not sure why this describes the `false` case only. I'd say it returns whether it should be supported, or if you want true/false:

```suggestion
// K8sNetworkPolicyEnabled returns true if cilium agent needs to support K8s NetworkPolicy, false otherwise.
```

**Review by qmonnet — CHANGES_REQUESTED** [review]
Looks OK, with some suggestions on grammar and phrasing for the comments and description of the flag - please see inline below. Thanks!

**Review by qmonnet — APPROVED** [review]
Looks good, thank you!

**Comment by tommyp1ckles** [comment]
TravicCI failure looks like: https://github.com/cilium/cilium/issues/23314

**Comment by qmonnet** [comment]
@ChengyuanLiCY Please, stop removing review assignees without stating a reason. I see @gandro already asked.

File: pkg/k8s/watchers/watcher.go
Comment (pchaigno) [inline pkg/k8s/watchers/watcher.go:452]: Should we enforce that `enable-policy=never` be set whenever this flag is passed? I'm a bit worried that someone might enable this and disable policy enforcement by mistake.

Comment (thorn3r) [reply pkg/k8s/watchers/watcher.go:452]: this seems like a good idea to me as well

Comment (ChengyuanLiCY) [reply pkg/k8s/watchers/watcher.go:452]: @pchaigno 
I mark this new flag as hidden as we discussed in the slack.

**Review by pchaigno — APPROVED** [review]
Thanks!

**Review by thorn3r — APPROVED** [review]

**Review by ldelossa — APPROVED** [review]

**Comment by pchaigno** [comment]
/test-1.16-4.19

**Comment by pchaigno** [comment]
k8s-1.16-kernel-4.19 and k8s-1.26-kernel-net-next hit known flake https://github.com/cilium/cilium/issues/22056. k8s-1.26-kernel-net-next also hit known flake https://github.com/cilium/cilium/issues/23637. Finally, k8s-1.16-kernel-4.19 failed with an existing flake (issue is new, but it's been happening for a while), https://github.com/cilium/cilium/issues/23833.

Given the above and the reviews, I'm merging.

**Comment by ChengyuanLiCY** [comment]
Thanks @pchaigno !

**Pull request merged**
